src/Interpreters/Session.cpp

```diff
@@ -441,15 +454,21 @@ ContextMutablePtr Session::makeSessionContext(const String & session_name_, std:
     prepared_client_info.reset();
 
     /// Set user information for the new context: current profiles, roles, access rights.
-    if (user_id && !new_session_context->getAccess()->tryGetUser())
-        new_session_context->setUser(*user_id);
+    ///  In the TCP handler, the quota key is received in "addendum packet" after context creation --> skip initialization.
+    if (!new_session_context->getAccess()->tryGetUser())
+        new_session_context->setUser(*user_id, res_client_info.interface != ClientInfo::Interface::TCP);
 
     /// Session context is ready.
     session_context = std::move(new_session_context);
     named_session = new_named_session;
     named_session_created = new_named_session_created;
     user = session_context->getUser();
 
+    session_tracker_handle = session_context->getSessionTracker().trackSession(
+        *user_id,
+        { session_name_ },
+        session_context->getSettingsRef().max_sessions_for_user);
```

Comment:
Can a user update the value of `max_sessions_for_user` in the `session_context` by executing a `SET` query, for example?

Reply:
Yes, but Behaviour will not be oblivious for user, 
SET set changes in current session, but it will not have effect, because `max_sessions_for_user` value is taken user profile, not session:

```sql
// in profile we have max_sessions_for_user = 0

Client_1 (user Alice):
SET max_sessions_for_user = 1 // changes value for client_1 only
SELECT value FROM  system.settings WHERE name="max_sessions_for_user" // return 1

Client 2 (user Alice):
SELECT value FROM  system.settings WHERE name="max_sessions_for_user" // return 0

Client 3 (user Alice): // connections count per user (3) already
SET max_sessions_for_user = 1 // active session willl continue.
```
So it looks like a bug to me. This bug can be applied to other settings from profile without contraints.

Reply:
This SQL looks like working:
```sql
ALTER SETTINGS PROFILE <dynamic_non_readonly_profile> SETTINGS max_sessions_for_user = 1
```
It will not close active sessions, but at least it will not allow to start new sessions.

I would certainly throw exception when trying to set up any settings taken from the profile for a session, but during quick looking into the code of how `SET `work and got no easy solution. I would create an separate issue for that and investigate the problem more thoughtful.

Reply:
> because max_sessions_for_user value is taken from user profile, not session

No, it's taken from `session_context`:
https://github.com/ClickHouse/ClickHouse/blob/d3e08a1e2a91cd4bef08b8270e5e68d05332a770/src/Interpreters/Session.cpp#L470

You probably mean that it's a new `session_context` that was just created and the user could not modify it. But it's not true for named http sessions because `NamedSessionsStorage` may return an existing session that was already in use, so the setting could be modified:
https://github.com/ClickHouse/ClickHouse/blob/d3e08a1e2a91cd4bef08b8270e5e68d05332a770/src/Interpreters/Session.cpp#L443-L444

Reply:
In the case of a named session, if `session_context `was already created, It is possible that `max_sessions_for_user `will have an old value:
https://github.com/ClickHouse/ClickHouse/blob/d3e08a1e2a91cd4bef08b8270e5e68d05332a770/src/Interpreters/Session.cpp#L458-L459

So, yes, it looks like a problem in this case:

```
ALTER SETTINGS PROFILE ALICE_PROFILE  max_sessions_for_user = 100
User  Alice started session_id_1 // session lifetime is 1000, max_session_per_user = 100
User  Alice started session_id_2 // session lifetime is 1000, max_session_per_user = 100
User  Alice started session_id_3 // session lifetime is 1000, max_session_per_user = 100
ALTER SETTINGS PROFILE ALICE_PROFILE  max_sessions_for_user = 1
User Alice finished session_id_3 // named session is not deleted and context remains valid with max_session_per_user = 100
User Alice started session_id_3 // existing context was reused with max_session_per_user = 100
```

Calling `new_session_context->setUser(*user_id)` always before reading max_sessions_for_user will fix the issue.
`setUser` is quite a complicated operation, do you know a better way to get the `max_sessions_for_user `value?

Reply:
> Calling new_session_context->setUser(*user_id) always before reading max_sessions_for_user will fix the issue.

And will break `SET` queries. 

> setUser is quite a complicated operation, do you know a better way to get the max_sessions_for_user value?

I don't know. We can take something from the implementation of `setUser`, maybe smth like `access->getDefaultProfileInfo()->settings.max_sessions_for_user` will work. Or we can save the value in `NamedSessionData` when creating a session. Or we can forbid changing the setting (see `Context::setSettings` and `Context::recalculateAccessRightsIfNeeded`)

Reply:
Thanks for the explanation.

Just pushed an update that should forbid using `SET max_sessions_for_user = 10` and several other parameters that shouldn't be updated by `SET query`  or  `SETTINGS` like
```bash
clickhouse-client -q "select value from system.settings where name='allow_settings_after_format_in_insert' SETTINGS max_sessions_for_user = 1"

// error
Received exception from server (version 23.7.1):
Code: 164. DB::Exception: Received from localhost:9000. DB::Exception: Setting max_sessions_for_user is not allowed to be set by query. (READONLY)
(query: select 1 SETTINGS max_sessions_for_user = 2)

```

`ALTER SETTINGS PROFILE` works fine.
```bash
clickhouse-client -q "create profile my_profile"
clickhouse-client -q "ALTER SETTINGS PROFILE 'my_profile' SETTINGS max_sessions_for_user = 2"
```

Reply:
I removed settings source restrictions from `SettingsConstraints.cpp` for other parameters, and left just for `max_sessions_for_user` just to not accidentally break something. This requester pool is already big.